--- a/cmake/zlib_external.cmake
+++ b/cmake/zlib_external.cmake
@@ -9,6 +9,6 @@ set(ZLIB_ENABLE_TESTS OFF CACHE INTERNAL "" FORCE)
 
 FetchContent_Declare(zlib
-                    GIT_REPOSITORY https://github.com/zlib-ng/zlib-ng
+                    GIT_REPOSITORY https://gitcode.com/gh_mirrors/zl/zlib-ng.git
                     GIT_TAG 2.0.6
                     USES_TERMINAL_DOWNLOAD TRUE)
 FetchContent_MakeAvailable(zlib)
